persistent/ring.py

```diff
@@ -0,0 +1,228 @@
+        Must not rely on object equality or object hashing, but only
+        identity or the `_p_oid`. Should be constant time.
+        """
+
+    def add(object):
+        """
+        Add the persistent object to the ring as most-recently used. When
+        an object is in the ring, the ring holds a strong reference to
+        it so it can be deactivated later by the pickle cache. Should
+        be constant time.
+
+        The object should not already be in the ring, but this is not necessarily
+        enforced.
+        """
+
+    def delete(object):
+        """
+        Remove the object from the ring if it is present. Returns a true
+        value if it was present and a false value otherwise. An ideal
+        implementation should be constant time, but linear time is
+        allowed.
+        """
+
+    def move_to_head(object):
+        """
+        Place the object as the most recently used object in the ring. The
+        object should already be in the ring, but this is not
+        necessarily enforced, and attempting to move an object that is
+        not in the ring has undefined consequences. An ideal
+        implementation should be constant time, but linear time is
+        allowed.
+        """
+
+    def delete_all(indexes_and_values):
+        """
+        Given a sequence of pairs (index, object), remove all of them from
+        the ring. This should be equivalent to calling :meth:`delete` for each
+        value, but allows for a more efficient bulk deletion process.
+
+        If the index and object pairs do not match with the actual state of the
+        ring, this operation is undefined.
+
+        Should be at least linear time (not quadratic).
+        """
+
+    def __iter__():
+        """
+        Iterate over each persistent object in the ring, in the order of least
+        recently used to most recently used. Mutating the ring while an iteration
+        is in progress has undefined consequences.
+        """
+
+from collections import deque
+
+@implementer(IRing)
+class _DequeRing(object):
+    """
+    A ring backed by the :class:`collections.deque` class. Operations
+    are a mix of constant and linear time.
+
+    It is available on all platforms.
+    """
+
+    __slots__ = ('ring', 'ring_oids')
+
+    def __init__(self):
+
+        self.ring = deque()
+        self.ring_oids = set()
+
+    def __len__(self):
+        return len(self.ring)
+
+    def __contains__(self, pobj):
+        return pobj._p_oid in self.ring_oids
+
+    def add(self, pobj):
+        self.ring.append(pobj)
+        self.ring_oids.add(pobj._p_oid)
+
+    def delete(self, pobj):
+        # Note that we do not use self.ring.remove() because that
+        # uses equality semantics and we don't want to call the persistent
+        # object's __eq__ method (which might wake it up just after we
+        # tried to ghost it)
+        i = 0 # Using a manual numeric counter instead of enumerate() is much faster on PyPy
```

Comment:
No, I haven't :( I can try to boil it down to a small test case for them.

In the meantime, that code shouldn't be used on PyPy anyway anymore, the CFFI implementation should be used, so I can go back to enumerate.